File: tcfunct.cpp
#include <iostream>
#include <memory>
#include <deque>
#include <algorithm>
#include <chrono>
 
template<typename T>
class Queue {
    struct Node {
        std::deque<T> front, back;
 
        Node(std::deque<T> f, std::deque<T> b) : front(f), back(b) {}
    };
    std::shared_ptr<Node> state;
 
public:
    Queue() : state(std::make_shared<Node>(std::deque<T>(), std::deque<T>())) {}
 
    Queue enqueue(T value) const {
        auto newBack = state->back;
        newBack.push_back(value);
        return Queue(std::make_shared<Node>(state->front, newBack));
    }
 
    std::pair<bool, T> dequeue() const {
        if (state->front.empty() && state->back.empty()) return {false, T()};
        auto newFront = state->front;
        if (newFront.empty()) {
            newFront = state->back;
            std::reverse(newFront.begin(), newFront.end());
        }
        T value = newFront.front();
        newFront.pop_front();
        return {true, value};
    }
 
    size_t getMemorySize() const {
        size_t size = sizeof(*state);
        size += sizeof(std::deque<T>) * 2;
        size += sizeof(T) * (state->front.size() + state->back.size());
        return size;
    }
 
private:
    Queue(std::shared_ptr<Node> node) : state(node) {}
};
 
template <typename Func>
void measureTime(const std::string& operation, Func&& func) {
    auto start = std::chrono::high_resolution_clock::now();
    func();
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> duration = end - start;
    std::cout << operation << " took " << duration.count() << " ms" << std::endl;
}
 
void testQueueOperations(size_t numElements) {
    Queue<int> queue;
 
    std::cout << "\nTesting " << numElements << " elements:\n";

    measureTime("Enqueue operation", [&]() {
        for (size_t i = 0; i < numElements; ++i) {
            queue = queue.enqueue(i);
        }
    });
    std::cout << "Memory usage after enqueue: " << queue.getMemorySize() << " bytes" << std::endl;

    measureTime("Dequeue operation", [&]() {
        for (size_t i = 0; i < numElements; ++i) {
            auto result = queue.dequeue();
        }
    });
}
 
int main() {
    testQueueOperations(10000);
    testQueueOperations(20000);
    testQueueOperations(30000);

    return 0;
}

File: tcfunct.hs
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import Control.DeepSeq (NFData, deepseq)
import GHC.Generics (Generic)

data Queue a = Queue [a] [a] deriving (Show, Generic)
instance NFData a => NFData (Queue a)

emptyQueue :: Queue a
emptyQueue = Queue [] []

enqueue :: a -> Queue a -> Queue a
enqueue x (Queue f r) = Queue f (x : r)

dequeue :: Queue a -> (Maybe a, Queue a)
dequeue (Queue [] []) = (Nothing, emptyQueue)
dequeue (Queue [] r) = dequeue (Queue (reverse r) [])
dequeue (Queue (x:xs) r) = (Just x, Queue xs r)

estimateMemorySize :: Queue a -> Int -> Int -> Int
estimateMemorySize (Queue front back) elementSize overhead =
    (length front + length back) * elementSize + overhead

measureTime :: String -> IO a -> IO a
measureTime description action = do
    start <- getCurrentTime
    result <- action
    end <- getCurrentTime
    putStrLn $ description ++ " took " ++ show (diffUTCTime end start)
    return result

forceEvaluation :: NFData a => a -> IO ()
forceEvaluation x = x `deepseq` return ()

testQueueOperations :: Int -> IO ()
testQueueOperations numElements = do
    let elementSize = 8 
    let overhead = 0   
    putStrLn $ "\nTesting " ++ show numElements ++ " elements:"
    measureTime "Enqueue operation" $ do
        let queue = foldl (\q x -> enqueue x q) emptyQueue [1..numElements]
        forceEvaluation queue
        putStrLn $ "Memory usage after enqueue: " 
                 ++ show (estimateMemorySize queue elementSize overhead) 
                 ++ " bytes"
    measureTime "Dequeue operation" $ do
        let queue = foldl (\q x -> enqueue x q) emptyQueue [1..numElements]
        let finalQueue = foldl (\(_, q) _ -> dequeue q) (Nothing, queue) [1..numElements]
        forceEvaluation (snd finalQueue)

main :: IO ()
main = do
    testQueueOperations 10000
    testQueueOperations 20000
    testQueueOperations 30000
